### PR TITLE
Feature: time varying body force

### PR DESCRIPTION
**Pull request by TzuYaoHuang**
I define a seperate function to apply time-varying body force in Flow solver. ( detailed description is in commits)

I also update .gitignore a bit to include more gfx and other stuffs to make it more comprehensive.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,11 +1,34 @@
...
 *.pvd
-Manifest.toml
-/docs/build/
-.vscode/
+*.pvt*
+
+# on supercomputer
+*.log
+slurm*
+*.sh*
```
Comment (b-fg) [inline .gitignore:33]: Bash scripts should not be excluded. For example, they are now used in the benchmark PR.
```suggestion
```

File: .gitignore [hunk 2]
```diff
@@ -1,11 +1,34 @@
...
+*.vtm
 *.pvd
-Manifest.toml
-/docs/build/
-.vscode/
+*.pvt*
+
+# on supercomputer
+*.log
+slurm*
```
Comment (b-fg) [inline .gitignore:32]: This looks like you are running cases in the same repo of the source code (maybe in `examples/`) and you are using the general `.gitignore` to ignore the files generated. While I think it is okay to add typical file extensions that surely won't be included in WaterLily, this system makes `.gitignore` an ever-growing bucket of different filenames. I would suggest having the WaterLily package in a directory, and an examples directory elsewhere. Then you can develop in the WaterLily directory normally, and in the other examples package mark WaterLily as a development package.

File: src/Flow.jl [hunk 3]
```diff
@@ -59,6 +59,15 @@ lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop (
     Φ[I] = ϕuP(j,CIj(j,CI(I,i),N[j]-2),CI(I,i),u,ϕ(i,CI(I,j),u)) -ν*∂(j,CI(I,i),u); r[I,i] += Φ[I]) over I ∈ slice(N,2,j,2)
 upperBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop r[I-δ(j,I),i] -= Φ[CIj(j,I,2)] over I ∈ slice(N,N[j],j,2)
 
+
+function applyBodyForce!(r::AbstractArray{T,Dv},t,g=(a,b)->0) where {T,Dv}
```
Comment (b-fg) [inline src/Flow.jl:63]: If `T,Dv` are not used within the function and there are no other `applyBofyForce!` function for which multiple dispatch would be required, there is not need to specify the type for `r`.
```suggestion
function applyBodyForce!(r,t,g=(a,b)->0)
```

File: src/Flow.jl [hunk 4]
```diff
@@ -85,9 +94,10 @@ struct Flow{D, T, Sf<:AbstractArray{T}, Vf<:AbstractArray{T}, Tf<:AbstractArray{
     U :: NTuple{D, T} # domain boundary values
     Δt:: Vector{T} # time step (stored in CPU memory)
     ν :: T # kinematic viscosity
+    g :: Function  # (possibly time-varying) body force field
     exitBC :: Bool # Convection exit
     perdir :: NTuple # direction of periodic direction
-    function Flow(N::NTuple{D}, U::NTuple{D}; f=Array, Δt=0.25, ν=0.,
+    function Flow(N::NTuple{D}, U::NTuple{D}; f=Array, Δt=0.25, ν=0., g::Function=(i,t)->0,
```
Comment (b-fg) [inline src/Flow.jl:100]: No need to specify `g` type either.
```suggestion
    function Flow(N::NTuple{D}, U::NTuple{D}; f=Array, Δt=0.25, ν=0., g=(i,t)->0,
```

File: src/util.jl [hunk 5]
```diff
@@ -147,6 +148,17 @@ in the other dimensions.
 function slice(dims::NTuple{N},i,j,low=1) where N
     CartesianIndices(ntuple( k-> k==j ? (i:i) : (low:dims[k]), N))
 end
+using Statistics
+"""
+    removeMean!(s)
+
+Remove the meam of scalar array `s` in physical domain from `s`.
```
Comment (b-fg) [inline src/util.jl:155]: `meam` typo.

File: src/util.jl [hunk 6]
```diff
@@ -147,6 +148,17 @@ in the other dimensions.
...
+This is potentially useful for pressure normalization.
+"""
+function removeMean!(s)
+    s̄ = Statistics.mean(@views s[inside(s)])
+    s .-= s̄
+end
```
Comment (b-fg) [inline src/util.jl:161]: Have you checked on allocations here (I see `@views` is used, but still)? Do we really need `Statistics` to compute the mean? This could be troublesome in GPUs.

File: src/Flow.jl [hunk 7]
```diff
@@ -59,6 +59,15 @@ lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop (
     Φ[I] = ϕuP(j,CIj(j,CI(I,i),N[j]-2),CI(I,i),u,ϕ(i,CI(I,j),u)) -ν*∂(j,CI(I,i),u); r[I,i] += Φ[I]) over I ∈ slice(N,2,j,2)
 upperBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop r[I-δ(j,I),i] -= Φ[CIj(j,I,2)] over I ∈ slice(N,N[j],j,2)
 
+
```
Comment (b-fg) [inline src/Flow.jl:62]: Docs missing

**Review by b-fg — CHANGES_REQUESTED** [review]
This is a good addition. I did a quick review on some basic stuff. Also I see that tests are failing at the moment, so might be worth looking into that.

File: .gitignore (hunk 2 above)
Comment (TzuYaoHuang) [reply .gitignore:32]: Sure! this is a very good suggestion. I think I will just make .gitignore ordered then.

File: src/Flow.jl (hunk 7 above)
Comment (TzuYaoHuang) [reply src/Flow.jl:62]: I am not sure why the change is included here. Maybe because I accidentally add a blank line here. This is the building blocks for Periodic BC

Comment (b-fg) [reply src/Flow.jl:62]: I just highlighted the line on top of your new function, where docs should be including describing the function header and what it does.

Comment (weymouth) [reply src/Flow.jl:64]: That functions isn't new to this PR. @TzuYaoHuang and/or @marinlauber should add docs for periodic BCs (including something on the main page mark-up) but not in this PR.

File: src/util.jl [hunk 8]
```diff
@@ -147,6 +148,16 @@ in the other dimensions.
 function slice(dims::NTuple{N},i,j,low=1) where N
     CartesianIndices(ntuple( k-> k==j ? (i:i) : (low:dims[k]), N))
 end
+"""
+    removeMean!(s)
+
+Remove the mean of scalar array `s` in physical domain from `s`.
+This is potentially useful for pressure normalization.
+"""
+function removeMean!(s)
```
Comment (weymouth) [inline src/util.jl:157]: Why do we need to add this?

File: src/util.jl [hunk 9]
```diff
@@ -25,6 +25,7 @@ Return a CartesianIndex of dimension `N` which is one at index `i` and zero else
 Return CartesianIndices range excluding a single layer of cells on all boundaries.
 """
 @inline inside(a::AbstractArray;buff=1) = CartesianIndices(map(ax->first(ax)+buff:last(ax)-buff,axes(a)))
+@inline inside(dims::NTuple{N}) where N = CartesianIndices(map(i->(2:i-1),dims))
```
Comment (weymouth) [inline src/util.jl:28]: Why do we need to add this?

File: src/WaterLily.jl [hunk 10]
```diff
@@ -54,16 +54,18 @@ struct Simulation
...
-        flow = Flow(dims,u_BC;uλ,Δt,ν,T,f=mem,perdir,exitBC)
+        flow = Flow(dims,u_BC;uλ,Δt,ν,g,T,f=mem,perdir,exitBC)
         measure!(flow,body;ϵ)
         new(U,L,ϵ,flow,body,MultiLevelPoisson(flow.p,flow.μ₀,flow.σ;perdir))
     end
 end
 
-time(sim::Simulation) = sum(sim.flow.Δt[1:end-1])
+time(flow::Flow) = sum(flow.Δt[1:end-1])
+time(sim::Simulation) = time(sim.flow)
```
Comment (weymouth) [inline src/WaterLily.jl:67]: We already have the function `sim_time`. Isn't this the same?

File: src/Flow.jl (hunk 7 above)
Comment (b-fg) [reply src/Flow.jl:64]: Ah my bad.

File: src/util.jl (hunk 6 above)
Comment (TzuYaoHuang) [reply src/util.jl]: Yes, @views won't allocate memory and you are right. Statistics.jl is not necessary.

File: src/Flow.jl [hunk 11]
```diff
@@ -59,6 +59,21 @@ lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop (
...
+Body force `g` input sould be function with arguments: index `a` and time `b`.
+The default of body force is null. 
+"""
+function applyBodyForce!(r,t,g=(a,b)->0)
+    N,n = size_u(r)
+    for i ∈ 1:n
+        gᵢ = g(i,t)
+        @loop r[I,i] += gᵢ over I ∈ inside(N)
+    end
+end
```
Comment (weymouth) [inline src/Flow.jl:75]: If you think `g` will be expensive, then I guess you could do
```suggestion
applyBodyForce!(a::Flow{N},t) where N = for i in 1:N
    gᵢ = a.g(i,t)
    a.f[:,i] .+= gᵢ
end
```
Broadcasting is plenty in this kind of case.

File: src/Flow.jl [hunk 12]
```diff
@@ -85,9 +100,10 @@ struct Flow{D, T, Sf<:AbstractArray{T}, Vf<:AbstractArray{T}, Tf<:AbstractArray{
     U :: NTuple{D, T} # domain boundary values
     Δt:: Vector{T} # time step (stored in CPU memory)
     ν :: T # kinematic viscosity
+    g :: Function  # (possibly time-varying) body force field
     exitBC :: Bool # Convection exit
     perdir :: NTuple # direction of periodic direction
-    function Flow(N::NTuple{D}, U::NTuple{D}; f=Array, Δt=0.25, ν=0.,
+    function Flow(N::NTuple{D}, U::NTuple{D}; f=Array, Δt=0.25, ν=0., g=(i,t)->0,
```
Comment (b-fg) [inline src/Flow.jl:106]: `g=(i,t)->0.`

File: src/Flow.jl [hunk 13]
```diff
@@ -128,11 +144,13 @@ and the `AbstractPoisson` pressure solver to project the velocity onto an incomp
     a.u⁰ .= a.u; scale_u!(a,0)
     # predictor u → u'
     conv_diff!(a.f,a.u⁰,a.σ,ν=a.ν,perdir=a.perdir)
+    applyBodyForce!(a.f,time(a),a.g)
```
Comment (b-fg) [inline src/Flow.jl:147]: Would it be more efficient to have a conditional here so that we avoid adding `0` when no body-force function is passed? Could you please check this @TzuYaoHuang?

**Review by weymouth — COMMENTED** [review]
It would also be nice to have an example of this capability in the example folder.

File: src/util.jl (hunk 8 above)
Comment (TzuYaoHuang) [reply src/util.jl:157]: This function is for pressure nomalization in runtests.jl. I can also leave it in runtest.jl. Just in case we are going to use it at some point in the future.

Comment (weymouth) [reply src/util.jl:157]: Even better would be to write a test that doesn't need a new function.

File: src/WaterLily.jl (hunk 10 above)
Comment (TzuYaoHuang) [reply src/WaterLily.jl:67]: `sim_time` is for dimensionless time calculation but also only works on the `::Simulation` type. What i am trying to do is to measure time with arguement of type `::Flow` since i need it in `mom_step!` but it cannot access the main simulation.

Comment (weymouth) [reply src/WaterLily.jl:67]: Sorry, I just mean `time` https://github.com/weymouth/WaterLily.jl/blob/6092807d7b76092977433b798db0b54df3bc2f36/src/WaterLily.jl#L66

Comment (weymouth) [reply src/WaterLily.jl:67]: Makes sense, but then you need to move it to `Flow.jl`, right?

File: src/Flow.jl (hunk 13 above)
Comment (TzuYaoHuang) [reply src/Flow.jl:147]: That is a good idea. I think I can add a flag in `Flow` struct and do sth generative function like in Periodic BC

File: src/util.jl (hunk 9 above)
Comment (TzuYaoHuang) [reply src/util.jl:28]: I added this line originally for the situation when gravity has the component in the periodic BC.
But yeah, add the value to all component including ghost cells does not matter.

File: src/Flow.jl (hunk 11 above)
Comment (TzuYaoHuang) [reply src/Flow.jl:75]: I will follow the idea of adding to all elements but I might still leave args the same since directly using `(a,t)` might limit the future usage. I can use `r::AbstractArray(T,Dv)` and use `Dv-1` to infer `N`.

File: src/WaterLily.jl (hunk 10 above)
Comment (TzuYaoHuang) [reply src/WaterLily.jl:67]: That's true!

**Comment by weymouth** [comment]
We also needs @marinlauber's comments since he's going to be using this for accelerating flows. Until it works for his use case, I don't think there is a point (until you merge the two-phase flow stuff).

**Comment by weymouth** [comment]
This example is great. This is exactly what @marinlauber needs. I'll check it out later, but I don't see how this works without updating the BC `U`.

**Comment by TzuYaoHuang** [comment]
In this case I just use periodic BC to bypass this, but I agree if this gonna to be used in real case, we need to update `U` also and probably can no longer stay uniform throughout the boundary.

File: src/Flow.jl [hunk 14]
```diff
@@ -59,6 +59,23 @@ lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop (
...
+
+This function apply body force `g` at time `t` to `r`, effectively the du/dt.
+Body force `g` input sould be function with arguments: index `a` and time `b`.
+The default of body force is null. Val{Bool} is applied to specify whether to run or not.
+"""
+function applyBodyForce!(r,t,::Val{true},g=(a,b)->0)
+    N,n = size_u(r)
+    for i ∈ 1:n
+        gᵢ = g(i,t)
+        r[CartesianIndices(N),i] .+= gᵢ
```
Comment (b-fg) [inline src/Flow.jl:73]:
```suggestion
        r[CartesianIndices(N),i] .+= g(i,t)
```

File: src/Flow.jl [hunk 15]
```diff
@@ -85,9 +102,11 @@ struct Flow{D, T, Sf<:AbstractArray{T}, Vf<:AbstractArray{T}, Tf<:AbstractArray{
     U :: NTuple{D, T} # domain boundary values
     Δt:: Vector{T} # time step (stored in CPU memory)
     ν :: T # kinematic viscosity
+    g :: Function  # (possibly time-varying) body force field
+    tagBF :: Bool # whether to switch on the body force
```
Comment (b-fg) [inline src/Flow.jl:106]: you do not really need `tagBF`. Consider de following workflow for multiple dispatch on `g`
```
foo(g, ::Val{nothing}) = "empty"
foo(g, ::Val) = "not empty"

g = nothing
foo(g, Val{g}()) # returns "empty"
g = x -> 0
foo(g, Val{g}()) # returns "not empty"
```

**Comment by weymouth** [comment]
> In this case I just use periodic BC to bypass this, but I agree if this gonna to be used in real case, we need to update `U` also and probably can no longer stay uniform throughout the boundary.

This is a good idea, and it's a valid use case for sure. For the other use case, the boundaries can stay uniform, but they need to be time varying and synced with the acceleration field. We want to prescribe `U = U(i,t)` and `g(i,t) = derivative(U(i,t),t)`. So maybe the thing to do is to avoid specifying the type of `g` in the `Flow` struct and use multiple dispatch as @b-fg suggested. Then we can set up another PR to do the same thing for `U` later.  What do you think?

File: src/Flow.jl (hunk 11 above)
Comment (weymouth) [reply src/Flow.jl]: What are you envisioning for the more general usage of the function? A general body force routine would need to be a function of `x`, not just `t`, so we should use `apply!(a.f,(i,x)->BodyForce(i,x,t))`, right?

Comment (weymouth) [reply src/Flow.jl]: Such a routine might be a good idea in the future (I _think_ that's a way we could get @b-fg 's SANS stuff integrated into the code) but this PR doesn't need to do everything everywhere all at once. ;-)

File: src/Flow.jl (hunk 15 above)
Comment (weymouth) [reply src/Flow.jl]: You only need `foo(::Nothing)="nothing"`.

**Comment by weymouth** [comment]
Ok - I've tried to clean this up a bit. Take a look. If you like this approach, please try to also clean up and minimize your test of the flow solver in `runtests.jl`. It's verbose and a bit hard to follow at the moment. Note that the old tests are already testing the default `g=nothing` condition, so we don't need to worry about that.

File: src/Flow.jl [hunk 16]
```diff
@@ -60,20 +59,15 @@ lowerBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop (
...
+This function apply uniform acceleration field `g` at time `t` to `r`.
+If `g ≠ nothing`, then `g(i,t)=dUᵢ/dt`.
 """
-function applyBodyForce!(r,t,::Val{true},g=(a,b)->0)
-    N,n = size_u(r)
-    for i ∈ 1:n
-        r[CartesianIndices(N),i] .+= g(i,t)
-    end
+accelerate!(r,t,g) = for i ∈ 1:last(size(r))
+    r[:,i] .= g(i,t)
```
Comment (TzuYaoHuang) [inline src/Flow.jl:68]: I think `r[CartesianIndices(N),i] .+= g(i,t)` is still necessary since the colon of `r[:,i]` can only get slice for one dimension, which cannot handle if dim(r)>2. 

Also I am wandering why the forcing vector `a.f` is overwritten. I guess it should be augmented. `.=` -> `.+=`

Comment (TzuYaoHuang) [inline src/Flow.jl:68]: I think `r[CartesianIndices(N),i] .+= g(i,t)` is still necessary since the colon of `r[:,i]` can only get slice for one dimension, which cannot handle if dim(r)>2. 

Also I am wandering why the forcing vector `r` is overwritten. I guess it should be augmented. That is, `.=` -> `.+=`
Because in multiphase flow I have to 1) a.f = \mu(gradU+gradU^T), 2) a.f ./= rho(f), 3) a.f .+= convection term 4) add gravity.
No matter the order of 3) or 4), we need to augment the body force term after finishing the calculation of diffusion term. step 1) and 2) cannot be combined together.

Comment (weymouth) [reply src/Flow.jl:68]: > I think `r[CartesianIndices(N),i]` is still necessary since the colon of `r[:,i]` can only get slice for one dimension, which cannot handle if dim(r)>2. (runtests.jl will yield the error)
Oops. I'm not sure how that didn't error when I checked it, but you're right.

> Also I am wandering why the forcing vector `r` is overwritten. I guess it should be augmented. That is, `.=` -> `.+=` Because in multiphase flow I have to 1) a.f = \mu(gradU+gradU^T), 2) a.f ./= rho(f), 3) a.f .+= convection term 4) add gravity. No matter the order of 3) or 4), we need to augment the body force term after finishing the calculation of diffusion term. step 1) and 2) cannot be combined together.

See line 100 of main.f90 in my old Fortran solver. The first step of every loop is setting up the background acceleration field.

Comment (weymouth) [reply src/Flow.jl]: Besides, the multi-phase stuff will need to be a whole different type, right? Not `Flow`, but `TwoPhase` or something. It'll need it's own `mom_step` function, etc. So there is no need to make this PR match exactly to what you're going to do in that function.

Comment (weymouth) [reply src/Flow.jl]: But, I'm happy going back to the `.+=` version. I'll push that up.

Comment (TzuYaoHuang) [reply src/Flow.jl]: The reason why in the fortran code we could specify domain acceleration first is because you have defined another variables `v` in line 79 in fvel.f90, However, in order to save the memory, we did not have any spare vector variable. as a result we need to divide `a.f` in situ when it comes to diffusion term. I appreciate for your understanding.

File: test/runtests.jl [hunk 17]
```diff
@@ -232,6 +240,41 @@ end
...
+        N=8
+        jerk = 0; gy = 1; sim = acceleratingFlow(N;jerk,gy,mem=f)
+        sim_step!(sim,1.0); timeExact = WaterLily.time(sim)
+        p = sim.flow.p |> Array
+        BC!(p,perdir=sim.flow.perdir)
+        pe = copy(sim.flow.p) |> Array; apply!((x)->sim.flow.g(2,timeExact)*(x[2]-N/2),pe)
+        BC!(pe,perdir=sim.flow.perdir)
+        @test WaterLily.L₂(p .- pe) < 5e-3 # the error due to accumulation of pressure solver tolerance
+    end
+end
```
Comment (weymouth) [inline test/runtests.jl:276]: I think this could be a lot smaller/cleaner. More like your example case. Then I think this will be ready to merge.

**Review by b-fg — APPROVED** [review]

File: test/runtests.jl (hunk 17 above)
Comment (TzuYaoHuang) [reply test/runtests.jl]: Aha I just make a commit on it. I hope it is clean enough. I delete the hydrostatic test case. and make a lot of comments.

**Comment by b-fg** [comment]
I would also like to see performance benchmarks on this. Maybe try the TGV already in benchmarks (should be like `g=nothing`) and compare it to TGV with `g=(i,t)->0.`.

**Comment by weymouth** [comment]
What's the behavior you expect to see?

**Comment by b-fg** [comment]
Hopefully that the multiple dispatch is not yielding some type instability making it slow. I think it should not be the case, but just to double check. Also doing a benchmark should become mandatory for PRs :) (it should now be very quick to do these comparisons with the new benchmark suit).

**Comment by weymouth** [comment]
I agree it should be the new standard. But I'm not clear on the approach.
Run the master and PR on the TGV, which should give identical results. Then
run using `g=0` which should be basically the same as well?

**Comment by b-fg** [comment]
Yeap, exactly!

File: test/runtests.jl [hunk 18]
```diff
@@ -232,6 +240,30 @@ end
...
+    # assuming gravitational scale is 1 and Fr is 1, U scale is Fr*√gL
+    UScale = √N  # this is also initial U
+    # constant jerk in x, zero acceleration in y
+    g(i,t) = i==1 ? t*jerk : 0
+    return WaterLily.Simulation(
+        (N,N), (UScale,0.), N; ν=0.001,g,Δt=0.001,perdir,T,mem
+    ),jerk
+end
+@testset "Flow.jl with increasing body force" begin
+    for f∈arrays
```
Comment (b-fg) [inline test/runtests.jl:254]:
```suggestion
    for f ∈ arrays
```

**Comment by weymouth** [comment]
I've installed git bash and just run the default benchmark suite. It finished instantly. Where is the json file?

**Comment by b-fg** [comment]
It should be in the same benchmark folder. Is this Windows you are running it?

**Comment by weymouth** [comment]
Yes, Windows. No file is being created, unfortunately.

**Comment by b-fg** [comment]
Okey it might be a problem with paths  since these programmed with Linux on mind only. Did the benchmark run at least? Also, is there an output error?

**Comment by weymouth** [comment]
I can't tell if it really run. No error message but no output and it runs instantly. 
image

**Comment by b-fg** [comment]
Ah, do you have juliaup installed? It's the only extra requirement to run benchmarks.

**Comment by b-fg** [comment]
It's also weird that it doesn't catch your Julia version (the field is empty in the info header)

**Comment by weymouth** [comment]
Yes, I have juliaup. There must be some issue with the script not finding a few of the execs...

**Comment by b-fg** [comment]
Found the bug, pushing the hotfix asap.

**Comment by weymouth** [comment]
Results! 
image
First WaterLily is master, second and third are this PR. First two are "tgv.jl" and the third is the tgv with `g=(i,t)->0`.

**Comment by weymouth** [comment]
I ran the benchmark and it seems ok to me. The timing for the master and this branch without the `g` argument match, and adding a trivial `g=(i,t)->0` slows the code down a few percent. 

However, the results on my laptop are extremely variable, so @TzuYaoHuang should probably post a final assessment using a Unix system.

**Comment by b-fg** [comment]
Sounds good, thanks. I guess you used the `-c` argument finally?

**Comment by weymouth** [comment]
I was using that argument the whole time. I think I got out of sync with the pushes to master. So I stashed my local changes synced up and unstashed and everything worked fine.

**Comment by TzuYaoHuang** [comment]
Benchmark result:

I tested three conditions as tested by Gabe: (1) gravity branch with default (`nothing`), (2) gravity branch with zero g (`g=(i,t)->0.`), and (3) the original master branch.

Generally speaking all the variations are within statistical error margin (on my linux laptop with only VSCODE open and I have run 2-3 times for each case to see the difference). However, I did observe that [A.] case 2 is constantly a bit better than case 1 when it comes to large grids, [B.] case 1 and case 3 is sorta similar, [C.] backend of CPUx6 seems to be more aligned to our prediction generally. Still again, everything is roughly with the error margin so no clear conclusion could be drawn IMHO.

[test3.log](https://github.com/weymouth/WaterLily.jl/files/13804344/test3.log)
[test1.log](https://github.com/weymouth/WaterLily.jl/files/13804345/test1.log)
[test2.log](https://github.com/weymouth/WaterLily.jl/files/13804346/test2.log)

**Comment by weymouth** [comment]
Adding an extra loop twice a time step can't be _faster_ than literally doing nothing. I'm guessing it's just noise. Which is good!

**Comment by weymouth** [comment]
@b-fg , do you want to double check our timing? Otherwise, this seems fine to merge.

**Comment by b-fg** [comment]
I will give it a proper look later today for a final check and then we
merge, but it should be all good. Thanks both for the work!
